### ISSUE TITLE
[Feature request] add other languages

### ISSUE THREAD

**Issue by darhma** [billthefarmer/gurgle#1]
Nice game, I was wondering if it was possible to add more languages for example integrating the possibility to take words from an external dictionary (like quickdic).

**Comment by billthefarmer** [comment]
I'm amazed! This app only got published today. It's perfectly possible to add other languages. However the app has two dictionaries, about 2,300 reasonably common words for the words to guess, and about 10,000 more obscure words for allowed guesses. If someone could come up with the equivalent in another language, it would be fairly easy to add the option. There are also nuances of spelling in different places in English which probably occur in other languages as well where they are spoken widely.

**Comment by darhma** [comment]
Thanks for the reply, I'm not a developer but I've seen the files of the "dictionaries" and it shouldn't be very difficult to create similar ones with other languages, more than anything else the problem is to find the words to insert in each "dictionary" and choose how to divide them. My native language is Italian, which from the point of view of spelling is much less problematic than English. Maybe in the future, time permitting and without promising anything, I will try to see if I can create a "dictionary" for Italian. If so, do you have any advice or recommendations in this regard?

**Comment by billthefarmer** [comment]
If you can find an Italian dictionary, it should be fairly easy using something like Linux text utilities like grep, sort, etc to pick out all the 5 letter words. Picking out a subset of more common words needs someone expert in the language.

**Comment by GrimPixel** [comment]
Hi, I would like to help
https://polyglotclub.com/wiki/Language/Multiple-languages/Culture/Internet-Dictionaries

I also think it's a good idea to display the meaning of the word when it's revealed.

**Comment by billthefarmer** [comment]
I had that in there to start off, but took it out for simplicity. I've put it back in again.

To create a version of this app for another language needs not just the dictionaries, but a translation of the strings and the help text. Then if you use the app on a phone set to that language it will just work in that language. See [How to create alternative resources](https://developer.android.com/guide/topics/resources/localization#creating-alternatives).

**Comment by vnourdin** [comment]
Hi :wave: 
Thanks for the great game :smile:
I play also a lot to [Lexica](https://github.com/lexica/lexica) which is free software too.
Maybe you can see how they did to manage multiple languages and both online and offline app dictionary like quickdic :shrug:

At first, I think there is no need to translate the app itself. Having multiple dictionaries would already bring more peoples to the game and thus more developers, which will be able to help to translate the app then :crossed_fingers:

**Comment by darhma** [comment]
Also lexica is a very nice game and I was thinking about lexica when I wrote the first post, about integrating the app with quickdic type dictionaries

**Comment by billthefarmer** [comment]
I had a look at the Lexica Italian dictionary, it has about 255K words in it and about 7K 5 letter words. If I used the whole 5 letter dictionary it would make the game far too difficult. It needs an expert, as I said above, to pick out the common words. This would apply to any language.

```shell
Bill@Cecelia MINGW64 /d/android/Gurgle/data (main)
$ wc dictionary.txt
 254854  254854 2329182 dictionary.txt

Bill@Cecelia MINGW64 /d/android/Gurgle/data (main)
$ egrep '^.{5}$' dictionary.txt | wc
   6980    6980   42550
```

**Comment by GrimPixel** [comment]
I would like to ask why it must be 5 letters. Isn't that limiting the game's potential?

**Comment by billthefarmer** [comment]
It doesn't, but then you would need four letter or six letter dictionaries.

**Comment by billthefarmer** [comment]
As an illustration of how difficult the app is without a common words dictionary I have taken all the 5 letter unaccented words out of the Lexica Italian dictionary and created a version with an Italian language option.

[Gurgle-debug.zip](https://github.com/billthefarmer/gurgle/files/7950481/Gurgle-debug.zip)

**Comment by darhma** [comment]
Thank you for creating a trial version for Italian. I've tried it in these two days making several games and although it is sometimes difficult to find the solution I think it is still pleasant to play, the problem I found is that sometimes the word to guess is not an existing word in Italian but it seems more like a Roman numeral: It happened to me with cccvi, cxcvi (of which I found the solution anyway) and in another case I found it impossible to find a solution because a green letter was m but the other letters left available were w h j k x and with this combination of letters it is impossible to create a word (unless it is a case like the others of a roman number to be created with m c x). 

In case if you let me have the used dictionary file I could review it to avoid this problem. It shouldn't be too complicated to search for words containing w j k x (which are very little used in Italian and a little less y) to check that there are only words that actually exist.

**Comment by darhma** [comment]
In order to simplify things, we could decide to exclude directly the words containing the letters w j k x (and maybe also y, which is rarely used anyway) from the Italian dictionary.

**Comment by darhma** [comment]
However, on average, apart from special cases, even without having separate dictionaries, it usually takes from 3 to 6/7 attempts to guess a word, which seems acceptable and not too frustrating.

**Comment by billthefarmer** [comment]
I thought I got all the Roman numerals, I checked, I missed some. I will check again before I release a version with Italian. I don't know why an Italian dictionary would have Roman numerals in it.

**Comment by billthefarmer** [comment]
The Italian dictionary is just`api/v1/language/it/dictionary.txt` from https://lexica.github.io/lexica/. I used `grep` to extract all the unaccented 5 character words like this...
```shell
$ grep '^[a-z]{5}$' dictionary.txt > words.txt
```
Then I noticed the Roman numerals...

**Comment by darhma** [comment]
I think you could already release an "official" version with the option of Italian, I don't see big problems, I took the liberty to spread this issue to others to have other opinions about it... I hope you don't mind. Once solved the problem of non-existent words I think there are no particular problems on other words maybe not so common. For example I guessed not so common words in 6/7 moves and other common words in more moves because in Italian if you have the last four green letters (ente) you can use g l m s d (and maybe some other option) to solve the puzzle, but these are all very common words in Italian so there is no way to diversify them

**Comment by billthefarmer** [comment]
OK, I think I have got all the Roman numbers, the Italian dictionary is here https://github.com/billthefarmer/gurgle/blob/main/src/main/java/org/billthefarmer/gurgle/Italian.java. These are the numbers I removed except for a few which look like genuine words.
```
cccii ccciv cccix cccli ccclv ccclx cccvi cccxc cccxi cccxl cccxv
cccxx cciii cclii ccliv cclix cclvi cclxi cclxv cclxx ccvii ccxci
ccxcv ccxii ccxiv ccxix ccxli ccxlv ccxvi ccxxi ccxxv ccxxx cdiii
cdlii cdliv cdlix cdlvi cdlxi cdlxv cdlxx cdvii cdxci cdxcv cdxii
cdxiv cdxix cdxli cdxlv cdxvi cdxxi cdxxv cdxxx cicli cliii climi
clivi clvii clxii clxiv clxix clxvi clxxi clxxv clxxx cmiii cmlii
cmliv cmlix cmlvi cmlxi cmlxv cmlxx cmvii cmxci cmxcv cmxii cmxiv
cmxix cmxli cmxlv cmxvi cmxxi cmxxv cmxxx cviii cxcii cxciv cxcix
cxcvi cxiii cxlii cxliv cxlix cxlvi cxvii cxxii cxxiv cxxix cxxvi
cxxxi cxxxv dccci dcccl dcccv dcccx dccii dcciv dccix dccli dcclv
dcclx dccvi dccxc dccxi dccxl dccxv dccxx dciii dclii dcliv dclix
dclvi dclxi dclxv dclxx dcvii dcxci dcxcv dcxii dcxiv dcxix dcxli
dcxlv dcxvi dcxxi dcxxv dcxxx dicci dilli dimmi dliii dlvii dlxii
dlxiv dlxix dlxvi dlxxi dlxxv dlxxx dviii dxcii dxciv dxcix dxcvi
dxiii dxlii dxliv dxlix dxlvi dxvii dxxii dxxiv dxxix dxxvi dxxxi
dxxxv licci lilli lviii lxiii lxvii lxxii lxxiv lxxix lxxvi lxxxi
lxxxv mccci mcccl mcccv mcccx mccii mcciv mccix mccli mcclv mcclx
mccvi mccxc mccxi mccxl mccxv mccxx mcdii mcdiv mcdix mcdli mcdlv
mcdlx mcdvi mcdxc mcdxi mcdxl mcdxv mcdxx mciii mclii mcliv mclix
mclvi mclxi mclxv mclxx mcmii mcmiv mcmix mcmli mcmlv mcmlx mcmvi
mcmxc mcmxi mcmxl mcmxv mcmxx mcvii mcxci mcxcv mcxii mcxiv mcxix
mcxli mcxlv mcxvi mcxxi mcxxv mcxxx mdccc mdcci mdccl mdccv mdccx
mdcii mdciv mdcix mdcli mdclv mdclx mdcvi mdcxc mdcxi mdcxl mdcxv
mdcxx mdiii mdlii mdliv mdlix mdlvi mdlxi mdlxv mdlxx mdvii mdxci
mdxcv mdxii mdxiv mdxix mdxli mdxlv mdxvi mdxxi mdxxv mdxxx mimmi
mliii mlvii mlxii mlxiv mlxix mlxvi mlxxi mlxxv mlxxx mmcii mmciv
mmcix mmcvi mmiii mmlii mmliv mmlix mmlvi mmlxi mmlxv mmlxx mmvii
mmxci mmxcv mmxii mmxiv mmxix mmxli mmxlv mmxvi mmxxi mmxxv mmxxx
mviii mxcii mxciv mxcix mxcvi mxiii mxlii mxliv mxlix mxlvi mxvii
mxxii mxxiv mxxix mxxvi mxxxi mxxxv villi xciii xcvii xliii xlvii
xviii xxiii xxvii xxxii xxxiv xxxix xxxvi
```

**Comment by darhma** [comment]
Good job, I had a quick look at the file and it seems to me that you have actually removed all the Roman numerals, I saw a few other words that I'm not too sure about like water which is an English word anyway and a few others. For now I'm pretty busy but tomorrow I should have some time available, in case I'll have a better look at it and if you think it's necessary I can also try to separate the dictionaries between common and less common words (I'm not an expert but at the university I took an exam in linguistics and I'm a "strong" reader so I think I have a wide enough vocabulary to be able to make some distinction), in case let me know.

**Comment by darhma** [comment]
I'll also see if I can do something about
> translation of the strings and the help text

**Comment by darhma** [comment]
Another thing I noticed is that if I start the app it loads the word to guess in English and if I change the dictionary (without restarting the game) the word to guess remains the English one but the dictionary used is the Italian one. I think that when you change the dictionary you should automatically restart the game, and in case it would be useful to make it so that when you open the app it automatically opens with the last dictionary used.

**Comment by darhma** [comment]
In case let me know as well if it is the case to eliminate or put among the possible words without being among those guessable swear words that could be offensive or simply swear words. For me it is not a problem to include swear words (very common in Italian) but for other users it could be...

**Comment by billthefarmer** [comment]
The debug version I uploaded was just the bare minimum to get it working. The thing with developing apps is like any other job, it takes much longer to get all the corner cases and nuances sorted than it does to just add the main functionality. I have released a new version with the Italian language option. It currently doesn't remember the language, but that could be in a later version as I have sorted out a Spanish and a Portuguese language, although the Portuguese language has far too many words in it.

I don't have an opinion on offensive words, if you look at any dictionary it will have some in it, it would be incomplete without.

**Comment by darhma** [comment]
I fully understand the problems in app development and my comment about language switching features was written more as a reminder than a request. Anyway, this morning I downloaded the dictionary file and started the word-by-word revision (I completed the letter d) creating a new file that contains archaic words, of very uncommon use or for which I didn't find satisfactory definitions on two different dictionaries I'm using for reference. I'm not sure if I should exclude particularly technical terms used only in specific fields (like chemistry, geology, medicine, etc). The work is quite long to do so it will take me some time to finish, but as said the app is already usable without too many problems. Another category of words that I'm not sure if I should leave or not are verbs to infinitive in truncated form (i.e. without the final letter e), for now I decided to leave them, in case later it should not be too complicated to remove them since they are all words ending with r (and in Italian there are not many others).
